src/Mod/Part/Gui/ViewProviderGridExtension.cpp

```diff
@@ -0,0 +1,513 @@
+    hGrp->Attach(this);
+}
+
+GridExtensionP::~GridExtensionP()
+{
+    Gui::coinRemoveAllChildren(GridRoot);
+    GridRoot->unref();
+
+    ParameterGrp::handle hGrp = App::GetApplication().GetParameterGroupByPath("User parameter:BaseApp/Preferences/Units");
+    hGrp->Detach(this);
+}
+
+double GridExtensionP::getGridSize() const
+{
+    return computedGridValue;
+}
+
+void GridExtensionP::getClosestGridPoint(double &x, double &y) const
+{
+    auto closestdim = [](double &dim, double gridValue) {
+        dim = dim / gridValue;
+        dim = dim < 0.0 ? ceil(dim - 0.5) : floor(dim + 0.5);
+        dim *= gridValue;
+    };
+
+    closestdim(x, computedGridValue);
+    closestdim(y, computedGridValue);
+
+    //Base::Console().Log("gridvalue=%f, (x,y)=(%f,%f)", computedGridValue, x, y);
+}
+
+bool GridExtensionP::checkCameraZoomChange(const Gui::View3DInventorViewer* viewer)
+{
+    float newCamMaxDimension = viewer->getMaxDimension();
+    if (fabs(newCamMaxDimension - camMaxDimension) > 0) { //ie if user zoomed.
+        camMaxDimension = newCamMaxDimension;
+        return true;
+    }
+
+    return false;
+}
+
+bool GridExtensionP::checkCameraTranslationChange(const Gui::View3DInventorViewer* viewer)
+{
+    //Then we check if user moved by more than 10% of camera dimension (must be after updating camera dimension).
+    SbVec3f newCamCenterOnSketch = viewer->getCenterPointOnFocalPlane();
+
+    if ((camCenterOnSketch - newCamCenterOnSketch).length() > 0.1 * camMaxDimension) {
+        camCenterOnSketch = newCamCenterOnSketch;
+        return true;
+    }
+
+    return false;
+}
+
+void GridExtensionP::computeGridSize(const Gui::View3DInventorViewer* viewer)
+{
+    if (!vp->GridAuto.getValue()) {
+        computedGridValue = vp->GridSize.getValue();
+        return;
+    }
+
+    short pixelWidth = -1;
+    short pixelHeight = -1;
+    viewer->getViewportRegion().getViewportSizePixels().getValue(pixelWidth, pixelHeight);
+    if (pixelWidth < 0 || pixelHeight < 0) {
+        computedGridValue = vp->GridSize.getValue();
+        return;
+    }
+
+    int numberOfLines = static_cast<int>(std::max(pixelWidth, pixelHeight)) / GridSizePixelThreshold;
+
+    double unitMultiplier = (unitsUserSchema == 2 || unitsUserSchema == 3) ? 25.4 : (unitsUserSchema == 5 || unitsUserSchema == 7) ? 304.8 : 1;
+
+    computedGridValue = vp->GridSize.getValue() * unitMultiplier * pow(GridNumberSubdivision, floor(log(camMaxDimension / unitMultiplier / numberOfLines) / log(GridNumberSubdivision)));
+
+    //cap the grid size
+    computedGridValue = std::max(computedGridValue, 0.000001);
+    computedGridValue = std::min(computedGridValue, 10000000.0);
+}
+
+void GridExtensionP::createGrid(bool cameraUpdate)
+{
+    Gui::MDIView* mdi = Gui::Application::Instance->editDocument()->getActiveView();
+    Gui::View3DInventorViewer* viewer = static_cast<Gui::View3DInventor*>(mdi)->getViewer();
```

Comment:
This static cast may be a bit dangerous (even if in normal case, that should always be OK).

Reply:
I can go along the dynamic_casting it, although I do not see a clear problem.